SMP: add expire notice on grid tile Thumbnails

sejas opened the pull request:
#### Proposed Changes

On #68965 A message was added for the Sites Management Page on list mode.
This PR introduces the same concept for Grid mode, a.k.a. Tiles.

* Add new component `SitesGridAction` to display messages in grid mode.
* Add new component `SitesGridActionRenew` to show the Renew CTA on expiring sites.
  * I'm using the same events used on `sites-plan-renew-nag.tsx`, I just added a new property `display_mode`.
  * I'm using the same icon on `sites-plan-renew-nag.tsx`.
  * I've added the plan name to the message.

#### Screenshots

1. `SitesGridActionRenew` on Coming Soon Site

<!--img width="437" alt="expire-coming-soon" src="https://user-images.githubusercontent.com/779993/197212454-bf0be360-f2e3-4ded-8039-1a0f78f6c33f.png" !-->
<!--img width="445" alt="redesign-coming-soon" src="https://user-images.githubusercontent.com/779993/197248543-ad46aae3-360f-4de5-88c0-4706a728a6ab.png"!-->
<!--img width="438" alt="Screenshot 2022-10-24 at 13 06 49" src="https://user-images.githubusercontent.com/779993/197522397-5e5a5075-234c-4564-a5be-d437d4ce48a7.png"!-->

<img width="439" alt="Screenshot 2022-10-25 at 15 06 19" src="https://user-images.githubusercontent.com/779993/197798616-d4e20d39-575c-4594-b313-49631d58766a.png">





2. `SitesGridActionRenew` on Public Site

<!--img width="435" alt="expire-public" src="https://user-images.githubusercontent.com/779993/197212485-a2c14a04-a715-42d2-96cc-0fdd31ce2e97.png"!-->

<!--img width="442" alt="redesign-public" src="https://user-images.githubusercontent.com/779993/197248563-11abb5ba-37c8-4cef-8b20-7c4bfa563315.png"!-->
<!--img width="450" alt="Screenshot 2022-10-24 at 13 06 30" src="https://user-images.githubusercontent.com/779993/197522444-4ccb2bc0-cc87-4458-8bfc-ceeb8b4ca565.png"!-->

<img width="452" alt="Screenshot 2022-10-25 at 15 07 10" src="https://user-images.githubusercontent.com/779993/197798127-ccfe637e-fee1-44a3-8c7d-a20eb18be617.png">


3. `SitesGridActionRenew` on Private Site

<!--img width="438" alt="expire-private" src="https://user-images.githubusercontent.com/779993/197212511-28c0ac51-52fb-403d-bef5-78da17a8647d.png"!-->
<!--img width="469" alt="redesign-private" src="https://user-images.githubusercontent.com/779993/197248584-ecbdea01-9ce1-4a76-a73f-74bfc3b72888.png"!-->

<!--img width="455" alt="Screenshot 2022-10-24 at 13 07 02" src="https://user-images.githubusercontent.com/779993/197522476-f79b2425-45c5-4883-a868-cc58c23cc88d.png"!-->

<img width="440" alt="Screenshot 2022-10-25 at 15 06 11" src="https://user-images.githubusercontent.com/779993/197798045-604d1aba-8885-40b5-bb64-fabcab85f8e6.png">



4. `SitesGridAction` with long text:

<!--img width="434" alt="expire-long-text" src="https://user-images.githubusercontent.com/779993/197212882-0b46d744-ac52-482c-8225-93498e146fcb.png"!-->


<!--img width="460" alt="redesign-long-sentence" src="https://user-images.githubusercontent.com/779993/197248824-a874d2de-34f7-4f11-b4ab-7679a0705eba.png"!-->

<!--img width="448" alt="Screenshot 2022-10-24 at 13 11 22" src="https://user-images.githubusercontent.com/779993/197522598-7fba79d4-a757-43a5-b644-4d832f6b4dae.png"!-->

<img width="455" alt="Screenshot 2022-10-25 at 15 15 03" src="https://user-images.githubusercontent.com/779993/197797903-c23e268e-cf3c-453e-b31a-4df735fa6d2a.png">


#### Testing Instructions

* Select an Atomic site and change the expiration date to the past using SA.
* Go to `/sites`
* Search for your site
* Observe a Notice appears over the Thumbnail.
* Only if you are the site owner you'll see a `Renew` CTA. Other users will see only the notice without the link.
* Repeat these steps for coming soon, private and public sites.

#### Pre-merge Checklist

- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [x] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [x] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?


- Closes https://github.com/Automattic/dotcom-forge/issues/1080
- Related to "#68965" which introduced the same concept for list mode

sejas committed:
SMP: create SitesGridAction component

sejas committed:
SMP: add SitesGridActionRenew styled component

sejas committed:
SMP: Show Expire Notice using SitesGridActionRenew on sites grid tile.

sejas committed:
SMP: change red color on action and remove dead code

sejas committed:
SMP: reuse PLAN_RENEW_NAG_IN_VIEW and ON_CLICK tracks events on SitesGridActionRenew Expired Site

sejas committed:
SMP: rename display_mode tracks property on SitesGridActionRenew

sejas committed:
SMP: Expire notice, change CTA color on SitesGridAction and fix widths

sejas committed:
SMP: make leadingAction optional on SitesGridTile

danielbachhuber commented on client/sites-dashboard/components/sites-grid-action-renew.tsx:
```diff
@@ -0,0 +1,52 @@
+interface SitesGridActionRenewProps {
+	site: SiteExcerptData;
+}
+export function SitesGridActionRenew( { site }: SitesGridActionRenewProps ) {
```
The other component uses these props:
```
{ isSiteOwner, plan, checkoutUrl }: PlanRenewProps
```

Should we follow a similar pattern, to avoid dependency on the large site object?

sejas replied:
I disagree with this approach in this particular case.

I've created two components:
- `SitesGridAction` one for the style/layout.
- `SitesGridActionRenew`, where the main goal is extracting what's needed from the site and passing the right props to `SitesGridAction`.

In your suggestion, I should move the site dependency to the parent's responsibility, which would be a fatter component. In your example, we are using the whole plan object. So it has the same degree of dependency.

danielbachhuber replied:
> I disagree with this approach in this particular case.

Sounds fine, we can go with your original approach.

danielbachhuber commented on client/sites-dashboard/components/sites-grid-action-renew.tsx:
```diff
@@ -0,0 +1,52 @@
+		>
+			{ createInterpolateElement(
+				/* translators: PlanName is the plan's product name, such as Pro or Business */
+				__( 'Your <PlanName /> plan has expired' ),
```
If the user isn't the site owner, "Your" probably doesn't make sense here.

Let's go with `__( '%s - Expired' )`, to follow the other component.

danielbachhuber commented on client/sites-dashboard/components/sites-plan-renew-nag.tsx:
```diff
@@ -7,8 +7,9 @@ import { sprintf } from '@wordpress/i18n';
-const PLAN_RENEW_NAG_ON_CLICK = 'calypso_sites_dashboard_plan_renew_nag_click';
+
+export const PLAN_RENEW_NAG_IN_VIEW = 'calypso_sites_dashboard_plan_renew_nag_inview';
+export const PLAN_RENEW_NAG_ON_CLICK = 'calypso_sites_dashboard_plan_renew_nag_click';
```
Would it make sense to abstract this to some common location, shared between both components?

sejas commented:
Redesign discussion: p1666361161617479-slack-C041RHH38NQ

danielbachhuber commented on client/sites-dashboard/components/sites-grid-action-renew.tsx:
```diff
@@ -0,0 +1,52 @@
+			ctaProps={
+				isSiteOwner && {
+					title: __( 'Renew' ),
+					href: `/checkout/${ site.slug }`,
```
I think this checkout link needs the plan too?

danielbachhuber commented on client/sites-dashboard/components/sites-plan-renew-nag.tsx:
```diff
@@ -87,7 +89,10 @@ export const PlanRenewNag = ( { isSiteOwner, plan, checkoutUrl }: PlanRenewProps
 				{ isSiteOwner && (
```
Can we remove the `<span>` from this string?

```
__( '<span>%s - Expired</span>' ),
```

There's no reason to include the HTML in the string here.

sejas replied:
done!

danielbachhuber commented on client/sites-dashboard/components/sites-plan-renew-nag.tsx:
```diff
@@ -7,8 +7,9 @@ import { sprintf } from '@wordpress/i18n';
 import { useInView } from 'react-intersection-observer';
-const PLAN_RENEW_NAG_IN_VIEW = 'calypso_sites_dashboard_plan_renew_nag_inview';
-const PLAN_RENEW_NAG_ON_CLICK = 'calypso_sites_dashboard_plan_renew_nag_click';
+
```
While we're touching this context, can we remove `react-intersection-observer` ?

sejas replied:
Sure !, done.

danielbachhuber commented on client/sites-dashboard/components/sites-grid-action.tsx:
```diff
@@ -0,0 +1,64 @@
+import { Gridicon } from '@automattic/components';
```
Nice! This component will be pretty sweet.

danielbachhuber replied:
@sejas Per @SaxonF's Slack comment (p1666382802271549/1666361161.617479-slack-C041RHH38NQ), is it worthwhile to import and reuse `calypso/components/notice` in this context?

sejas replied:
> There should be an existing notice component we can rely on so that we aren’t rebuilding/duplicating anything, it also ensures everything is consistent like border radius etc

If the goal of re-using the component is to "ensure everything is consistent like border radius, etc", then this is how it would look using the calypso `Notice` component:

<img width="443" alt="Screenshot 2022-10-24 at 17 34 27" src="https://user-images.githubusercontent.com/779993/197578736-93e00ae9-b12d-45c4-b7b7-77c1774c4b45.png">

It's an easy change, and I have the diff saved on my local machine. @danielbachhuber let me know what you prefer.

danielbachhuber replied:
@sejas Let's go with the `Notice` component so we aren't reinventing the wheel. Sorry I didn't catch that initially!

zaguiini replied:
That's more visually consistent with overall Calypso, yes I like that. The only thing that I kind of don't like but am "OK" with is the fact that the notice has square borders while the container has them rounded 😅  Not a big deal but it does look a bit inconsistent in this page's scope.

sejas replied:
I've updated the code. It's a medium refactor.
Here is the Before > After for each site status:

1. `SitesGridActionRenew` on Coming Soon Site

| Before | After |
|---|---|
| <img width="438" alt="Screenshot 2022-10-24 at 13 06 49" src="https://user-images.githubusercontent.com/779993/197522397-5e5a5075-234c-4564-a5be-d437d4ce48a7.png">  | <img width="439" alt="Screenshot 2022-10-25 at 15 06 19" src="https://user-images.githubusercontent.com/779993/197798616-d4e20d39-575c-4594-b313-49631d58766a.png"> |


2. `SitesGridActionRenew` on Public Site

| Before | After |
|---|---|
| <img width="450" alt="Screenshot 2022-10-24 at 13 06 30" src="https://user-images.githubusercontent.com/779993/197522444-4ccb2bc0-cc87-4458-8bfc-ceeb8b4ca565.png"> | <img width="452" alt="Screenshot 2022-10-25 at 15 07 10" src="https://user-images.githubusercontent.com/779993/197798127-ccfe637e-fee1-44a3-8c7d-a20eb18be617.png"> |

3. `SitesGridActionRenew` on Private Site

| Before | After |
|---|---|
|<img width="455" alt="Screenshot 2022-10-24 at 13 07 02" src="https://user-images.githubusercontent.com/779993/197522476-f79b2425-45c5-4883-a868-cc58c23cc88d.png">| <img width="440" alt="Screenshot 2022-10-25 at 15 06 11" src="https://user-images.githubusercontent.com/779993/197798045-604d1aba-8885-40b5-bb64-fabcab85f8e6.png"> |



4. `SitesGridAction` with long text:

| Before | After |
|---|---|
|<img width="448" alt="Screenshot 2022-10-24 at 13 11 22" src="https://user-images.githubusercontent.com/779993/197522598-7fba79d4-a757-43a5-b644-4d832f6b4dae.png">| <img width="455" alt="Screenshot 2022-10-25 at 15 15 03" src="https://user-images.githubusercontent.com/779993/197797903-c23e268e-cf3c-453e-b31a-4df735fa6d2a.png"> |

danielbachhuber reviewed (commented):
Great start! Left some comments to discuss.

sejas committed:
SMP: SitesGridAction redesign applying suggestion  p1666361161617479-slack-C041RHH38NQ

sejas committed:
SMP: change wording on Expire, from 'Your <PlanName /> plan has expired' to 'Pro - Expired'

sejas committed:
SMP: add plan to checkout url to SitesGridActionRenew

sejas committed:
SMP: move Expire event names to common object PLAN_RENEW_EVENT_NAMES

sejas committed:
SMP: refactor PlanRenewNag to use our useInView

sejas commented:
@danielbachhuber, thanks for reviewing.
I've applied your suggestions except for removing `Site` from props. Happy to debate about it.
I've also used the new redesign. 👌

sejas commented:
In this commit https://github.com/Automattic/wp-calypso/pull/69339/commits/fda77d5166ee5f47b536a76430d7fe75e0077e42 we've reduced the wording from
`Your Pro plan has expired` to `Pro - Expired`.

@danielbachhuber Suggested this change so the invited users are not confused by the word `Your`, and we could also share the translation with the list mode.

@zaguiini has suggested making it more descriptive: `Plan expired: Pro`. In the list mode, we have the column header that adds context about the Plan.

What do you think if we just say:
`Plan Pro has expired` 

Or we could use two sentences one for owners and another for invited users.
As a follow-up, we could add something like: `Contact the site's owner to renew the plan.`, maybe linking to the user's profile?

zaguiini commented:
Looking very very good! How about adding a slight drop shadow to the notice so it's clearer when there's little contrast between it and the thumbnail?

danielbachhuber commented:
I think `Business Plan expired.` reads a little better.

sejas committed:
SMP: fix useInView usage passing the ref and wrapping the callback

sejas committed:
SMP: change Expire copy on grid mode

sejas committed:
SMP: add box shadow on Expire grid action

sejas commented:
@danielbachhuber @zaguiini I've [changed the copy](https://github.com/Automattic/wp-calypso/pull/69339/commits/e28f9c9bd3b7a79f5e963212a6c75ee6e26d2323) and added [a box shadow](https://github.com/Automattic/wp-calypso/pull/69339/commits/35332479dfa64af7dfe5638dcb9c394d24fd149b).
I've updated the screenshots to match the latest design with the shadow.

sejas committed:
Merge branch 'trunk' of github.com:Automattic/wp-calypso into add/smp-expire-notice-cta-on-grid-tile

danielbachhuber commented on client/sites-dashboard/components/sites-grid-action.tsx:
```diff
@@ -0,0 +1,66 @@
+	marginRight: 16,
+	'.gridicon': {
+		color: 'var(--color-site-grid-action-icon)',
+		marginRight: 16,
```
It seems to look a little better with a tighter `4px` margin. What do you think?

## Before

<img width="900" alt="image" src="https://user-images.githubusercontent.com/36432/197549561-5d515f10-4d23-4af9-b54a-e33591f1482b.png">


## After

<img width="907" alt="image" src="https://user-images.githubusercontent.com/36432/197549464-9b088c07-c918-4fc1-9358-99132df37130.png">

sejas replied:
Done! https://github.com/Automattic/wp-calypso/pull/69339/commits/031e437d5598f0f59d81b2b0dfdaecf3757d2eff
The original space was pulled from the designs. But I also prefer the short spacing 👌.

danielbachhuber commented on client/sites-dashboard/components/sites-grid-action.tsx:
```diff
@@ -0,0 +1,66 @@
+	boxShadow: '0 0 4px var(--studio-gray-60)',
+} );
+
+const CONTENT = css( {
```
Is there a reason you went with `css()` instead of `styled.div()`?

zaguiini replied:
Yeah. In this case it actually makes more sense to use `styled.div` since you're immediately using the class name in a single component with the same tag, so I don't see a good reason to go with the `css` factory function.

sejas replied:
I think it's a taste preference.
I've found that using CSS classNames offers more benefits than styled-components:

**Pros**
1. I can still read the HTML without any abstraction, so it's more semantic to me.
2. I don't need to check if the component is just an HTML element or a full react component. I know `a` is just a link. But if I see `Cta` or `RenewLink`, maybe it's a full react component with more logic inside. In short files, I can see that it's indeed a styled component, but if the file is large, I would need to scroll to ensure it's just a styled component.

What do you think? @danielbachhuber @zaguiini 

| StyledComponent  |
|---|
| <img width="848" alt="Screenshot 2022-10-25 at 10 33 41" src="https://user-images.githubusercontent.com/779993/197738719-abf998a5-4b15-4e56-8a53-a6460cc79bc9.png">  |

| css |
|---|
|<img width="859" alt="Screenshot 2022-10-25 at 10 42 06" src="https://user-images.githubusercontent.com/779993/197740504-c2969853-4a2c-46e8-af00-b2a02bd65f0c.png">|

In the CSS case, I can confirm that the only external component is `GridIcon` at a glance.
It would be even better with the [`css prop`](https://emotion.sh/docs/css-prop).

danielbachhuber replied:
@sejas Personally, I think we should use `styled.div` because that's the pattern we've settled on elsewhere.

zaguiini replied:
That makes sense too, @sejas. I don't have strong opinions as both works. However, I'd like to be consistent and settle for one of the implementations... Definitely a follow-up, but that's something to keep an eye on.

sejas replied:
Thank you @zaguiini ! I see that our current code base uses 80% styled-components and 20% CSS.
So I can change this PR in favor of styled components and bring back the conversation in the future. I like the idea of never stopping to look for new alternatives to improve our code.

Thank you both! 🙌

danielbachhuber commented on client/sites-dashboard/utils.ts:
```diff
@@ -41,3 +41,8 @@ export const MEDIA_QUERIES = {
 	wide: '@media screen and ( min-width: 1280px )',
 };
+
+export const PLAN_RENEW_EVENT_NAMES = {
```
To make this more specific, can we rename this to `PLAN_RENEW_NAG_EVENT_NAMES`

sejas replied:
Done! https://github.com/Automattic/wp-calypso/pull/69339/commits/8bb72145c3414d912a0549a0938d32af5a01e56d

danielbachhuber reviewed (commented):
Getting closer! Left a few comments.

sejas committed:
SMP: reduce icon space on SitesGridAction

zaguiini commented on client/sites-dashboard/components/sites-grid-action-renew.tsx:
```diff
@@ -0,0 +1,54 @@
+interface SitesGridActionRenewProps {
+	site: SiteExcerptData;
+}
+export function SitesGridActionRenew( { site }: SitesGridActionRenewProps ) {
```
We're referencing `site.plan?.product_slug` multiple times throughout this component. Each distinct `site.plan?.product_slug` usage adds more code to the final bundle as this has to be transpiled to JS. How about extracting it to a variable (i.e. `planSlug`) and then referencing this variable?

sejas replied:
Done! https://github.com/Automattic/wp-calypso/pull/69339/commits/4371c0cd250e9f83a3bee43e68f6da8419026b07

I preferred to call it productSlug, so it's more consistent with the rest of the code.

zaguiini reviewed (commented):
Let's not forget to use logical properties in CSS-in-JS, otherwise, it'll look funky in RTL languages!

p-jackson commented:
> With a long sentence to see how it fits here

😻

sejas committed:
SMP: abstract productSlug on SitesGridActionRenew

sejas committed:
SMP: rename PLAN_RENEW_NAG_EVENT_NAMES

zaguiini commented on client/sites-dashboard/components/sites-grid-action.tsx:
```diff
@@ -0,0 +1,66 @@
+} );
+
+const CTA = css( {
+	'--color-link': '#111619',
```
What's the point of this CSS variable here? It's a simple `a` element, right? Is it used by any child?

sejas replied:
Not used anymore.

zaguiini commented on client/sites-dashboard/components/sites-grid-action.tsx:
```diff
@@ -0,0 +1,66 @@
+				{ children }
+			</div>
+			{ ctaProps && (
+				<a className={ CTA } { ...ctaProps } children={ ctaProps.children || ctaProps.title } />
```
Not technically wrong but definitely unusual seeing the `children` as a prop instead of the actual wrapped content by the element 😅 It might definitely confuse some peers, so let's try changing this:

```suggestion
				<a className={ CTA } { ...ctaProps }>
				    { ctaProps.children || ctaProps.title }
				</a>
```

sejas replied:
It's always good to remember that `children` is just one more prop :P.

Applied the change 👍 , thanks for the suggestion.

zaguiini commented on client/sites-dashboard/components/sites-grid-item.tsx:
```diff
@@ -95,6 +97,7 @@ export const SitesGridItem = memo( ( { site }: SitesGridItemProps ) => {
 					/>
 				</ThumbnailLink>
 			}
+			leadingAction={ site.plan?.expired && <SitesGridActionRenew site={ site } /> }
```
We don't need to add more props to the tile component, especially in this case.

We can use JSX fragments inside the `leading` prop, like:

```tsx
<SitesGridTile
    leading={
        <>
            <ThumbnailLink {...} />
            { site.plan?.expired && <SitesGridActionRenew site={ site } /> }
        </>
     }
/>

sejas replied:
Love it!, I didn't see it.

sejas committed:
SitesGridActionRenew: refactor in favor of Notice

sejas committed:
SMP: undo SitesGridItem changes

zaguiini commented on client/sites-dashboard/components/sites-grid-item.tsx:
```diff
@@ -84,16 +86,19 @@ export const SitesGridItem = memo( ( { site }: SitesGridItemProps ) => {
+							sizesAttr={ SIZES_ATTR }
+						/>
+					</ThumbnailLink>
+					site.plan?.expired && <SitesGridActionRenew site={ site } />
```
We need to wrap this line with curly braces as this is an expression
```suggestion
					{ site.plan?.expired && <SitesGridActionRenew site={ site } /> }
```

![image](https://user-images.githubusercontent.com/26530524/197836805-442a3f88-b2f9-443f-af6a-0a40d76aab04.png)

😬

zaguiini commented on client/sites-dashboard/components/sites-grid-action-renew.tsx:
```diff
@@ -0,0 +1,80 @@
+
+	return (
+		<Container>
+			<Notice icon="notice" showDismiss={ false }>
```
How about setting an error status in this notice, so it stands out more to the user?

| Right now | Proposed |
| ---------- | ---------- |
| ![image](https://user-images.githubusercontent.com/26530524/197837696-8f7c2ce3-ed09-4b1f-80e0-29de1c7a635b.png) | ![image](https://user-images.githubusercontent.com/26530524/197837516-d787f093-39a9-4db0-8b3a-3f26972bb4a5.png) |

What is needed:

```suggestion
			<Notice status="is-error" icon="notice" showDismiss={ false }>
```

danielbachhuber committed:
Make sure the expression is evaluated

Co-authored-by: Luis Felipe Zaguini <26530524+zaguiini@users.noreply.github.com>

danielbachhuber committed:
Make the notice stand out more prominently

Co-authored-by: Luis Felipe Zaguini <26530524+zaguiini@users.noreply.github.com>

danielbachhuber committed:
Merge branch 'trunk' into add/smp-expire-notice-cta-on-grid-tile

danielbachhuber reviewed (approved):
Looks and works great. Nice work on this, @sejas 🦾

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7593625

Thank you @sejas for including a screenshot in the description! This is really helpful for our translators.
